Logging UnitTests failing intermittantly

Occasionally, I'll see the unit tests fail with one or two test failures in the `/logging/*` space.

Here's an example of one of such failures:
```
2018-02-22T21:07:17.1653665Z com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests > testLoggerMessageIsRetainedWhenReportingException STANDARD_ERROR
2018-02-22T21:07:17.1654670Z     AI: INFO 22-02-2018 21:07, 10(Test worker): Configuration file has been successfully found as resource
2018-02-22T21:07:19.3888085Z 
2018-02-22T21:07:19.3889202Z com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests > testLoggerMessageIsRetainedWhenReportingException FAILED
2018-02-22T21:07:19.3889986Z     java.lang.AssertionError: expected:<1> but was:<2>
2018-02-22T21:07:19.3890754Z         at org.junit.Assert.fail(Assert.java:88)
2018-02-22T21:07:19.3891250Z         at org.junit.Assert.failNotEquals(Assert.java:834)
2018-02-22T21:07:19.3891790Z         at org.junit.Assert.assertEquals(Assert.java:645)
2018-02-22T21:07:19.3892306Z         at org.junit.Assert.assertEquals(Assert.java:631)
2018-02-22T21:07:19.3893328Z         at com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests.testLoggerMessageIsRetainedWhenReportingException(ApplicationInsightsAppenderTests.java:71)
```

This assertion is checking how many telemetry items were received by the mocked channel. Locally, I've only seen failures with something along the lines of `java.lang.AssertionError: expected:<1> but was:<`**`0`**`>`, so I thought it was a timing issue.

To test this hypothesis I added a 1 s. sleep before the assertion. Here was the result:
```
com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests > testLoggerMessageIsRetainedWhenReportingException FAILED
    java.lang.AssertionError: expected:<1> but was:<55>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests.testLoggerMessageIsRetainedWhenReportingException(ApplicationInsightsAppenderTests.java:76)

com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests > testAppenderSendsGivenEvent FAILED
    java.lang.AssertionError: expected:<1> but was:<40>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppenderTests.testAppenderSendsGivenEvent(ApplicationInsightsAppenderTests.java:67)
```
Very interesting.

I don't think this is something wrong with the code, but the test setup. I'm adding a JUnit `@Rule` to dump the collected telemetry when there's a failure. Hopefully that will reveal the root cause.

--- a/.gitignore
+++ b/.gitignore
@@ -95,3 +95,6 @@ $RECYCLE.BIN/
 /Backend/OrderService/build
 /Backend/OrderService/.idea
 Backend/OrderService/OrderService.iml
+
+# VS Code Workspace settings
+.vscode/settings.json

--- a/logging/log4j1_2/appenderSrc/test/java/com/microsoft/applicationinsights/log4j/v1_2/ApplicationInsightsAppenderTests.java
+++ b/logging/log4j1_2/appenderSrc/test/java/com/microsoft/applicationinsights/log4j/v1_2/ApplicationInsightsAppenderTests.java
@@ -22,10 +22,18 @@
 package com.microsoft.applicationinsights.log4j.v1_2;
 
 import java.util.List;
-import org.junit.*;
+import java.util.concurrent.TimeUnit;
 
+import org.junit.*;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+
+import com.google.common.collect.Maps;
 import com.microsoft.applicationinsights.internal.shared.LogChannelMockVerifier;
+import com.microsoft.applicationinsights.telemetry.ExceptionTelemetry;
 import com.microsoft.applicationinsights.telemetry.Telemetry;
+import com.microsoft.applicationinsights.telemetry.TraceTelemetry;
+
 import org.apache.log4j.LogManager;
 import org.apache.log4j.Logger;
 
@@ -42,7 +50,30 @@ public class ApplicationInsightsAppenderTests {
     @Before
     public void setup() {
         setMockTelemetryChannelToAIAppender();
+        Assert.assertTrue("MockTelemetryChannel not clear....", LogChannelMockVerifier.INSTANCE.getTelemetryCollection().isEmpty());
     }
+
+    @Rule
+    public TestWatcher watcher = new TestWatcher() {
+        @Override
+        protected void failed(Throwable e, Description description) {
+            System.out.println("Failure detected. Printing found telemetry:");
+            List<Telemetry> telemetries = LogChannelMockVerifier.INSTANCE.getTelemetryCollection();
+            for (Telemetry t : telemetries) {
+                String tname = t.getClass().getSimpleName();
+                String tinfo = tname;
+                if (t instanceof TraceTelemetry) {
+                    TraceTelemetry trct = (TraceTelemetry)t;
+                    tinfo = String.format("%s{message='%s'}", tname, trct.getMessage());
+                } else if (t instanceof ExceptionTelemetry) {
+                    ExceptionTelemetry ext = (ExceptionTelemetry)t;
+                    tinfo = String.format("%s{%s[message='%s']}", tname, ext.getException().getClass().getSimpleName(), ext.getException().getMessage());
+                }
+                System.out.println(" - "+tinfo);
+            }
+            System.out.printf("%d total telemetries collected.%n", telemetries.size());
+        }
+    };
 
     // endregion Initialization & cleanup
 
@@ -57,17 +88,20 @@ public class ApplicationInsightsAppenderTests {
     }
 
     @Test
-    public void testAppenderSendsGivenEvent() {
+    public void testAppenderSendsGivenEvent() throws Exception {
         Logger logger = LogManager.getRootLogger();
         logger.trace("New event!");
+        TimeUnit.SECONDS.sleep(1);
 
         Assert.assertEquals(1, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
     }
 
     @Test
-    public void testLoggerMessageIsRetainedWhenReportingException() {
+    public void testLoggerMessageIsRetainedWhenReportingException() throws Exception {
         Logger logger = LogManager.getRootLogger();
         logger.error("This is an exception", new Exception("Fake Exception"));
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertEquals(1, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().containsKey("Logger Message"));
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().get("Logger Message").equals("This is an exception"));

--- a/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
+++ b/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
@@ -26,11 +26,20 @@
     <appender name="test" class="com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppender">
         <param name="instrumentationKey" value="c9341531-05ac-4d8c-972e-36e97601d5ff" />
         <layout class="org.apache.log4j.PatternLayout">
-            <param name="ConversionPattern" value="%-5p %c{1} - %m%n"/>
+            <param name="ConversionPattern" value="%-5p %c - %m%n"/>
         </layout>
     </appender>
+    <appender name="console" class="org.apache.log4j.ConsoleAppender">
+        <layout class="org.apache.log4j.PatternLayout">
+            <param name="ConversionPattern" value="%-5p %c - %m%n"/>
+        </layout>
+    </appender>
+    <logger name="com.microsoft.applicationinsights.core.dependencies.http">
+        <level value="off" />
+    </logger>
     <root>
-        <priority value ="trace" />
+        <level value ="TRACE" />
         <appender-ref ref="test" />
+        <appender-ref ref="console" />
     </root>
 </log4j:configuration>

--- a/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
+++ b/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
@@ -34,9 +34,12 @@
             <param name="ConversionPattern" value="%-5p %c - %m%n"/>
         </layout>
     </appender>
+
+    <!-- suppresses embedded http logger -->
     <logger name="com.microsoft.applicationinsights.core.dependencies.http">
         <level value="off" />
     </logger>
+    
     <root>
         <level value ="TRACE" />
         <appender-ref ref="test" />

--- a/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
+++ b/logging/log4j1_2/appenderSrc/test/resources/log4j.xml
@@ -35,7 +35,7 @@
         </layout>
     </appender>
 
-    <!-- suppresses embedded http logger -->
+    <!-- suppresses repackaged HttpClient logger -->
     <logger name="com.microsoft.applicationinsights.core.dependencies.http">
         <level value="off" />
     </logger>

--- a/logging/log4j2/appenderSrc/test/java/com/microsoft/applicationinsights/log4j/v2/ApplicationInsightsAppenderTests.java
+++ b/logging/log4j2/appenderSrc/test/java/com/microsoft/applicationinsights/log4j/v2/ApplicationInsightsAppenderTests.java
@@ -23,6 +23,7 @@ package com.microsoft.applicationinsights.log4j.v2;
 
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
 import com.microsoft.applicationinsights.internal.shared.LogChannelMockVerifier;
 import com.microsoft.applicationinsights.telemetry.Telemetry;
@@ -91,24 +92,30 @@ public class ApplicationInsightsAppenderTests {
     }
 
     @Test
-    public void testAppenderDoesNotAppendLogsBelowFilterThreshold() {
+    public void testAppenderDoesNotAppendLogsBelowFilterThreshold() throws Exception {
         Logger logger = LogManager.getRootLogger();
         logger.trace("New event!");
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertEquals(0, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
     }
 
     @Test
-    public void testAppenderAppendsItemOnAndAboveSetThreshold() {
+    public void testAppenderAppendsItemOnAndAboveSetThreshold() throws Exception {
         Logger logger = LogManager.getRootLogger();
         logger.error("This should be appended!");
         logger.fatal("This is fatal is reported");
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertEquals(2, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
     }
 
     @Test
-    public void testLoggerMessageIsRetainedWhenReportingException() {
+    public void testLoggerMessageIsRetainedWhenReportingException() throws Exception {
         Logger logger = LogManager.getRootLogger();
         logger.error("This is an exception", new Exception("Fake Exception"));
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertEquals(1, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().containsKey("Logger Message"));
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().get("Logger Message").equals("This is an exception"));

--- a/logging/logback/appenderSrc/test/java/com/microsoft/applicationinsights/logback/ApplicationInsightsAppenderTests.java
+++ b/logging/logback/appenderSrc/test/java/com/microsoft/applicationinsights/logback/ApplicationInsightsAppenderTests.java
@@ -26,6 +26,7 @@ import org.junit.Before;
 import org.junit.Test;
 
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 
 import com.microsoft.applicationinsights.internal.shared.LogChannelMockVerifier;
 import com.microsoft.applicationinsights.telemetry.Telemetry;
@@ -82,19 +83,23 @@ public class ApplicationInsightsAppenderTests {
     }
 
     @Test
-    public void testLoggerMessageIsRetainedWhenReportingException() {
+    public void testLoggerMessageIsRetainedWhenReportingException() throws Exception {
         Logger logger = (Logger) LoggerFactory.getLogger("root");
         logger.error("This is an exception", new Exception("Fake Exception"));
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertEquals(1, LogChannelMockVerifier.INSTANCE.getTelemetryCollection().size());
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().containsKey("Logger Message"));
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().get("Logger Message").equals("This is an exception"));
     }
 
     @Test
-    public void testMDCPropertiesAreBeingSetAsCustomDimensions() {
+    public void testMDCPropertiesAreBeingSetAsCustomDimensions() throws Exception {
         Logger logger = (Logger) LoggerFactory.getLogger("root");
         MDC.put("key", "value");
         logger.error("This is an exception", new Exception("Fake Exception"));
+        TimeUnit.SECONDS.sleep(1);
+
         Assert.assertTrue(LogChannelMockVerifier.INSTANCE.getTelemetryCollection().get(0).getProperties().get("key").equals("value"));
     }
 

--- a/logging/logging.gradle
+++ b/logging/logging.gradle
@@ -21,7 +21,7 @@
 
 apply from: "$buildScriptsDir/common-java.gradle"
 apply from: "$buildScriptsDir/publishing.gradle"
-apply plugin: "findbugs"
+apply plugin: 'findbugs'
 
 def buildSrcDir = new File(project.buildDir, "src")
 def commonSrcDir = new File(project.parent.projectDir, "common/src")
@@ -56,6 +56,7 @@ sourceSets {
     }
 }
 
+findbugs.sourceSets = [sourceSets.main]
 
 // region Publishing properties
 

